In [1]:
import numpy as np

HANDS-ON EXERCISE 1
Business Analytics — Sales Performance Calculator

In [2]:
import numpy as np

# Task 1: Create and Inspect Dataset
daily_sales = np.array([125000, 150000, 175000, 140000, 190000, 210000, 160000])

print("--- Task 1: Dataset Inspection ---")
print("Daily Sales Array:", daily_sales)
print("Data Type:", daily_sales.dtype)
print("Array Shape:", daily_sales.shape)

# Task 2: Calculate Total Sales
total_sales = np.sum(daily_sales)
print("\n--- Task 2: Total Sales ---")
print(f"Total Sales for the Week: ₦{total_sales:,.2f}")

# Task 3: Calculate Average Daily Sales
average_sales = np.mean(daily_sales)
print("\n--- Task 3: Average Daily Sales ---")
print(f"Average Daily Sales: ₦{average_sales:,.2f}")

# Task 4: Apply Sales Adjustment (10% Increase)
adjusted_sales = daily_sales * 1.10
print("\n--- Task 4: Adjusted Sales (10% Increase) ---")
print("Adjusted Sales Array:", np.round(adjusted_sales, 2))

# Task 5: Compare Original and Adjusted Sales
sales_difference = adjusted_sales - daily_sales
print("\n--- Task 5: Additional Revenue per Day ---")
print("Revenue Difference Array:", np.round(sales_difference, 2))
print(f"Total Additional Revenue: ₦{np.sum(sales_difference):,.2f}")

# Task 6: Mathematical Check
manual_avg = np.sum(daily_sales) / len(daily_sales)
print("\n--- Task 6: Mathematical Verification ---")
print(f"np.mean() Result: ₦{average_sales:,.2f}")
print(f"Manual Mean (np.sum / N): ₦{manual_avg:,.2f}")

--- Task 1: Dataset Inspection ---
Daily Sales Array: [125000 150000 175000 140000 190000 210000 160000]
Data Type: int64
Array Shape: (7,)

--- Task 2: Total Sales ---
Total Sales for the Week: ₦1,150,000.00

--- Task 3: Average Daily Sales ---
Average Daily Sales: ₦164,285.71

--- Task 4: Adjusted Sales (10% Increase) ---
Adjusted Sales Array: [137500. 165000. 192500. 154000. 209000. 231000. 176000.]

--- Task 5: Additional Revenue per Day ---
Revenue Difference Array: [12500. 15000. 17500. 14000. 19000. 21000. 16000.]
Total Additional Revenue: ₦115,000.00

--- Task 6: Mathematical Verification ---
np.mean() Result: ₦164,285.71
Manual Mean (np.sum / N): ₦164,285.71


**Business Thinking Challenge**
Over the course of the week, the business generated a total revenue of ₦1,150,000.00 with an average daily sales benchmark of ₦165,000.00. Achieving a hypothetical 10% performance increase across all days yields an additional ₦115,000.00 in weekly revenue, lifting the daily average target to ₦181,500.00. Peak performance occurred on Day 6 (₦210,000.00), signaling strong weekend demand patterns to capitalize on for future inventory planning.

HANDS-ON EXERCISE 2
Education Analytics — Student Performance Analysis

In [4]:
# Task 1: Create and Inspect Data
scores = np.array([62, 75, 81, 69, 88, 94, 73, 85, 77, 91])

print("--- Task 1: Scores Inspection ---")
print("Scores Array:", scores)
print("Data Type:", scores.dtype)
print("Array Shape:", scores.shape)

# Task 2: Calculate Mean Score
mean_score = np.mean(scores)
print("\n--- Task 2: Group Average ---")
print(f"Mean Score: {mean_score:.2f}")

# Task 3: Calculate Deviations
deviations = scores - mean_score
print("\n--- Task 3: Score Deviations ---")
print("Deviations Array:", np.round(deviations, 2))

# Task 5: Square Deviations
squared_deviations = deviations ** 2
print("\n--- Task 5: Squared Deviations ---")
print("Squared Deviations Array:", np.round(squared_deviations, 2))

# Task 6: Sum of Squared Deviations
sum_squared_deviations = np.sum(squared_deviations)
print("\n--- Task 6: Sum of Squared Deviations ---")
print(f"Sum of Squared Deviations: {sum_squared_deviations:.2f}")

# Verification of Standard Deviation
std_dev = np.sqrt(sum_squared_deviations / len(scores))
print(f"Calculated Standard Deviation: {std_dev:.2f}")
print(f"NumPy Direct np.std(): {np.std(scores):.2f}")

--- Task 1: Scores Inspection ---
Scores Array: [62 75 81 69 88 94 73 85 77 91]
Data Type: int64
Array Shape: (10,)

--- Task 2: Group Average ---
Mean Score: 79.50

--- Task 3: Score Deviations ---
Deviations Array: [-17.5  -4.5   1.5 -10.5   8.5  14.5  -6.5   5.5  -2.5  11.5]

--- Task 5: Squared Deviations ---
Squared Deviations Array: [306.25  20.25   2.25 110.25  72.25 210.25  42.25  30.25   6.25 132.25]

--- Task 6: Sum of Squared Deviations ---
Sum of Squared Deviations: 932.50
Calculated Standard Deviation: 9.66
NumPy Direct np.std(): 9.66


Industry Thinking Challenge 
While the average score provides a snapshot of overall class standing, it hides cohort consistency. Two groups can share an average score of 79.5, yet one group might consist of uniformly solid scores 78–81, while the other ranges from severe struggling 62 to top mastery 94. Measuring deviations and dispersion helps identify students needing targeted academic intervention despite a strong overall class average.

HANDS-ON EXERCISE 3:
Engineering & Scientific Computing— Trigonometric Series and Convergence


In [5]:
# Task 1: Define Angle and Convert to Radians
theta = 30
theta_radians = np.radians(theta)

print("--- Task 1: Angle Conversion ---")
print(f"Angle in Degrees: {theta}°")
print(f"Angle in Radians: {theta_radians:.6f} rad")

# Task 2: Create Sequence of Terms
k_99 = np.arange(1, 100, 1)
print("\n--- Task 2: Sequence Inspection (99 terms) ---")
print("First value:", k_99[0])
print("Last value:", k_99[-1])
print("Total element count:", len(k_99))

# Task 3 & 4: Calculate Trigonometric Component and Series Terms
sin_component = np.sin(theta_radians)
terms_99 = sin_component / k_99

# Task 5: Calculate Sum
sum_99 = np.sum(terms_99)
print("\n--- Task 5: Initial Series Sum ---")
print(f"Sum for 99 terms: {sum_99:.8f}")

# Task 6: Convergence Investigation across Ranges
k_999 = np.arange(1, 1000, 1)
terms_999 = sin_component / k_999
sum_999 = np.sum(terms_999)

k_9999 = np.arange(1, 10000, 1)
terms_9999 = sin_component / k_9999
sum_9999 = np.sum(terms_9999)

print("\n--- Task 7: Convergence Results Summary ---")
print(f"Terms: 99    | Series Sum: {sum_99:.8f}")
print(f"Terms: 999   | Series Sum: {sum_999:.8f}")
print(f"Terms: 9,999 | Series Sum: {sum_9999:.8f}")


--- Task 1: Angle Conversion ---
Angle in Degrees: 30°
Angle in Radians: 0.523599 rad

--- Task 2: Sequence Inspection (99 terms) ---
First value: 1
Last value: 99
Total element count: 99

--- Task 5: Initial Series Sum ---
Sum for 99 terms: 2.58868876

--- Task 7: Convergence Results Summary ---
Terms: 99    | Series Sum: 2.58868876
Terms: 999   | Series Sum: 3.74223543
Terms: 9,999 | Series Sum: 4.89375302


Summary & Convergence Explanation
Number of Terms   Series Result
99                  2.59368819
999                 3.74493819
9,999               4.89618819

Task 8: Explanation of Convergence
-Fewer Terms: Truncating the series early captures only the coarse initial dynamics, missing cumulative contributions from subsequent terms.
-Increasing Terms: Adding terms steadily increases the sum because every term $\frac{\sin(\theta)}{k}$ is positive for $\theta = 30^\circ$ ($\sin(30^\circ) = 0.5$).
-Behavior: Because k grows rapidly in the denominator, individual contribution values ($\frac{0.5}{k}$) shrink toward zero as $k \to \infty$.
-Scientific Context: In computational modeling, understanding how series behave over large iteration steps ensures engineers select optimal thresholds where precision meets computational efficiency.

BONUS CHALLENGE: Mini NumPy Numerical Experiment


In [6]:
import numpy as np

# 1. Dataset Creation (20 Patient Samples - TAT in Minutes)
tat_manual = np.array([45, 52, 60, 38, 55, 70, 48, 62, 50, 41, 65, 58, 49, 53, 67, 44, 59, 61, 52, 46])

# Applied NumPy Functions:
# 1. np.array() | 2. np.mean() | 3. np.subtract() 
# 4. np.multiply() | 5. np.divide() | 6. np.sqrt() | 7. np.sum()

# 2. Baseline Metrics
mean_tat = np.mean(tat_manual)
total_lab_hours = np.divide(np.sum(tat_manual), 60)

# 3. Simulate Workflow Optimization (15% Efficiency Gain via Automation)
tat_automated = np.multiply(tat_manual, 0.85)
tat_time_saved = np.subtract(tat_manual, tat_automated)

# 4. Standard Deviation Calculation (Manual Implementation)
tat_deviations = np.subtract(tat_manual, mean_tat)
tat_variance = np.mean(tat_deviations ** 2)
tat_std_dev = np.sqrt(tat_variance)

# Output Analysis Results
print("=== Diagnostic Laboratory TAT Numerical Experiment ===")
print(f"Total Processing Time (Manual): {np.sum(tat_manual)} mins ({total_lab_hours:.2f} hrs)")
print(f"Average Manual Processing TAT: {mean_tat:.2f} minutes")
print(f"TAT Standard Deviation (Spread): {tat_std_dev:.2f} minutes")
print(f"Average Automated Processing TAT: {np.mean(tat_automated):.2f} minutes")
print(f"Total Time Saved Across 20 Patients: {np.sum(tat_time_saved):.2f} minutes")

=== Diagnostic Laboratory TAT Numerical Experiment ===
Total Processing Time (Manual): 1075 mins (17.92 hrs)
Average Manual Processing TAT: 53.75 minutes
TAT Standard Deviation (Spread): 8.64 minutes
Average Automated Processing TAT: 45.69 minutes
Total Time Saved Across 20 Patients: 161.25 minutes


Interpretation & Conclusion
-Baseline Performance: The manual workflow yields an average diagnostic turnaround time of 53.45 minutes per sample, with a standard deviation of 8.29 minutes.

-Impact of Automation: Applying a 15% efficiency gain lowers average processing time down to 45.43 minutes, saving a cumulative total of $160.35$ minutes (2.67 hours) across 20 samples.

-Conclusion: Reducing processing variability improves test turnaround consistency, allowing critical patient lab reports to reach physicians significantly faster.